src/toil/cwl/cwltoil.py

```diff
@@ -109,6 +115,79 @@
     "ResolveIndirect",
 )
 
+# What exit code do we need to bail with if we or any of the local jobs that
+# parse workflow files see an unsupported feature?
+CWL_UNSUPPORTED_REQUIREMENT_EXIT_CODE = 33
+
+# And what error will make the worker exit with that code
+CWL_UNSUPPORTED_REQUIREMENT_EXCEPTION = UnsupportedRequirement
+
+
+# Customized CWL utilities
+
+def visit_top_cwl_class(
+    rec: MutableMapping,
+    classes: Iterable[str],
+    op: Callable[[MutableMapping], Any]
+) -> None:
+    """
+    Apply the given operation to all top-level CWL objects with the given named CWL class.
+    Like cwltool's visit_class but doesn't look inside any object visited.
+    """
+    if isinstance(rec, MutableMapping):
+        if rec.get("class", None) in classes:
+            # This is one of the classes requested
+            # So process it
+            op(rec)
+        else:
+            # Look inside it instead
+            for key in rec:
+                visit_top_cwl_class(rec[key], classes, op)
+    elif isinstance(rec, MutableSequence):
+        # This item is actually a list of things, so look at all of them.
+        for key in rec:
+            visit_top_cwl_class(key, classes, op)
+
+# TODO: we need some type constraints here that I don't know how to spell.
```

Comment:
Check out https://github.com/common-workflow-language/cwltool/blob/dc5381e04fa2a9d18de5c78e0f8c2907919983ee/cwltool/utils.py#L85 and friends

Reply:
I guess `CWLObjectType` could be used in place of `MutableMapping`, but this comment really refers to how the `Any` returned from `op_down` needs to be the same `Any` accepted by `op_up`, for a given call to this function. I don't know if Python's type system supports that kind of constraint.

Reply:
Try `OpType= TypeVar('OpType')` and use that

Reply:
OK, I'll define a TypeVar for the down operation return type.